File: apps/ios/Wordocious/Sources/HeadToHeadService.swift
import Foundation
import Supabase

/// All-time head-to-head record + minimal opponent profile for the VS
/// intro/header/result UI — ports apps/web/lib/head-to-head.ts.
struct HeadToHeadRecord: Equatable {
    var myWins = 0
    var theirWins = 0
    var draws = 0
}

/// Minimal public profile bits needed by the VS intro/header/result UI.
struct VsProfile: Equatable {
    let username: String
    let avatarUrl: String?
    let level: Int
}

enum HeadToHeadService {
    private struct MatchRow: Decodable {
        let player1_id: String
        let player2_id: String?
        let winner_id: String?
    }

    /// All-time head-to-head record between two players, counted from the
    /// `matches` table (rows where the two ids occupy player1/player2 in
    /// either order). A draw is a VS row (player2_id set) with no winner_id.
    static func fetchHeadToHead(myId: String, opponentId: String) async -> HeadToHeadRecord {
        let rows: [MatchRow] = (try? await AuthService.shared.client.from("matches")
            .select("player1_id, player2_id, winner_id")
            .or("and(player1_id.eq.\(myId),player2_id.eq.\(opponentId)),and(player1_id.eq.\(opponentId),player2_id.eq.\(myId))")
            .limit(1000)
            .execute().value) ?? []

        var record = HeadToHeadRecord()
        for row in rows {
            if row.winner_id == myId { record.myWins += 1 }
            else if row.winner_id == opponentId { record.theirWins += 1 }
            else if row.player2_id != nil { record.draws += 1 }
        }
        return record
    }

    private struct ProfileBits: Decodable {
        let username: String?
        let avatar_url: String?
        let level: Int?
    }

    /// Minimal public profile for the VS UI — ports fetchVsProfile.
    static func fetchVsProfile(userId: String) async -> VsProfile? {
        let rows: [ProfileBits] = (try? await AuthService.shared.client.from("profiles")
            .select("username, avatar_url, level")
            .eq("id", value: userId)
            .limit(1)
            .execute().value) ?? []
        guard let row = rows.first else { return nil }
        return VsProfile(username: row.username ?? "Player", avatarUrl: row.avatar_url, level: row.level ?? 1)
    }

    /// "You lead 3–1" / "<name> leads 2–1" / "Tied 1–1" / "First meeting!" —
    /// ports match-intro.tsx headToHeadLine.
    static func headToHeadLine(opponentName: String, _ h2h: HeadToHeadRecord) -> String {
        if h2h.myWins == 0 && h2h.theirWins == 0 && h2h.draws == 0 { return "First meeting!" }
        if h2h.myWins > h2h.theirWins { return "You lead \(h2h.myWins)–\(h2h.theirWins)" }
        if h2h.theirWins > h2h.myWins { return "\(opponentName) leads \(h2h.theirWins)–\(h2h.myWins)" }
        return "Tied \(h2h.myWins)–\(h2h.theirWins)"
    }
}

File: apps/android/app/src/main/kotlin/com/wordocious/app/data/HeadToHeadService.kt
package com.wordocious.app.data

import io.github.jan.supabase.postgrest.postgrest
import io.github.jan.supabase.postgrest.query.Columns
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * All-time head-to-head record + minimal opponent profile for the VS
 * intro/header/result UI — Android port of web lib/head-to-head.ts.
 */
object HeadToHeadService {
    private val client get() = SupabaseConfig.client

    data class HeadToHeadRecord(val myWins: Int, val theirWins: Int, val draws: Int)

    /** Minimal public profile bits needed by the VS intro/header/result UI. */
    @Serializable
    data class VsProfile(
        val username: String? = null,
        @SerialName("avatar_url") val avatarUrl: String? = null,
        val level: Int? = null,
    ) {
        val displayName: String get() = username?.takeIf { it.isNotBlank() } ?: "Player"
    }

    @Serializable
    private data class MatchRow(
        @SerialName("player1_id") val player1Id: String? = null,
        @SerialName("player2_id") val player2Id: String? = null,
        @SerialName("winner_id") val winnerId: String? = null,
    )

    /** Web match-intro headToHeadLine — identical copy in all four states. */
    fun headToHeadLine(opponentName: String, h2h: HeadToHeadRecord): String = when {
        h2h.myWins == 0 && h2h.theirWins == 0 && h2h.draws == 0 -> "First meeting!"
        h2h.myWins > h2h.theirWins -> "You lead ${h2h.myWins}–${h2h.theirWins}"
        h2h.theirWins > h2h.myWins -> "$opponentName leads ${h2h.theirWins}–${h2h.myWins}"
        else -> "Tied ${h2h.myWins}–${h2h.theirWins}"
    }

    /**
     * All-time head-to-head between two players, counted from the `matches`
     * table (rows where the two ids occupy player1/player2 in either order).
     * A draw is a VS row (player2_id set) with no winner_id.
     */
    suspend fun fetchHeadToHead(myId: String, opponentId: String): HeadToHeadRecord = runCatching {
        val rows = client.postgrest["matches"]
            .select(Columns.raw("player1_id, player2_id, winner_id")) {
                filter {
                    or {
                        and { eq("player1_id", myId); eq("player2_id", opponentId) }
                        and { eq("player1_id", opponentId); eq("player2_id", myId) }
                    }
                }
                limit(1000)
            }
            .decodeList<MatchRow>()
        var myWins = 0; var theirWins = 0; var draws = 0
        for (row in rows) {
            when {
                row.winnerId == myId -> myWins++
                row.winnerId == opponentId -> theirWins++
                row.player2Id != null -> draws++
            }
        }
        HeadToHeadRecord(myWins, theirWins, draws)
    }.getOrDefault(HeadToHeadRecord(0, 0, 0))

    /** Opponent's public profile (username/avatar/level) — null if not found. */
    suspend fun fetchVsProfile(userId: String): VsProfile? = runCatching {
        client.postgrest["profiles"]
            .select(Columns.raw("username, avatar_url, level")) {
                filter { eq("id", userId) }
                limit(1)
            }
            .decodeSingleOrNull<VsProfile>()
    }.getOrNull()
}
